source/distributed/solution_transfer.cc

```diff
@@ -206,73 +201,81 @@ namespace parallel
+          cell->get_interpolated_dof_values(*(*cit_input), *it_output);
         }
+
+      Utilities::pack(dofvalues, data, /*allow_compression=*/false);
     }
 
 
+
     template <int dim, typename VectorType, typename DoFHandlerType>
     void
     SolutionTransfer<dim, VectorType, DoFHandlerType>::unpack_callback(
       const typename Triangulation<dim, DoFHandlerType::space_dimension>::
         cell_iterator &cell_,
       const typename Triangulation<dim, DoFHandlerType::space_dimension>::
         CellStatus /*status*/,
-      const void *               data,
-      std::vector<VectorType *> &all_out)
+      const boost::iterator_range<std::vector<char>::const_iterator> data_range,
+      std::vector<VectorType *> &                                    all_out)
     {
       typename DoFHandlerType::cell_iterator cell(*cell_, dof_handler);
 
       const unsigned int dofs_per_cell = cell->get_fe().dofs_per_cell;
-      ::dealii::Vector<typename VectorType::value_type> dofvalues(
-        dofs_per_cell);
-      const typename VectorType::value_type *data_store =
-        reinterpret_cast<const typename VectorType::value_type *>(data);
 
-      for (typename std::vector<VectorType *>::iterator it = all_out.begin();
-           it != all_out.end();
-           ++it)
+      std::vector<::dealii::Vector<typename VectorType::value_type>> dofvalues =
+        Utilities::unpack<
+          std::vector<::dealii::Vector<typename VectorType::value_type>>>(
+          data_range.begin(), data_range.end(), /*allow_compression=*/false);
+
+      // check if sizes match
+      Assert(dofvalues.size() == all_out.size(), ExcInternalError());
+
+      // check if we have enough dofs provided by the FE object
+      // to interpolate the transferred data correctly
+      for (auto it_dofvalues = dofvalues.begin();
+           it_dofvalues != dofvalues.end();
+           ++it_dofvalues)
+        Assert(dofs_per_cell <= it_dofvalues->size(),
+               ExcMessage(
+                 "The transferred data was packed on less dofs than the"
+                 "currently registered FE object assigned to the DoFHandler."));
+
+      // distribute data for each registered vector on mesh
+      auto it_input  = dofvalues.begin();
+      auto it_output = all_out.begin();
+      for (; it_input != dofvalues.end(); ++it_input, ++it_output)
         {
-          std::memcpy(&dofvalues(0),
-                      data_store,
-                      sizeof(typename VectorType::value_type) * dofs_per_cell);
-          cell->set_dof_values_by_interpolation(dofvalues, *(*it));
-          data_store += dofs_per_cell;
+          // Adjust size of vector to the order of current FE object
+          // associated with the registered DofHandler.
+          // Fixes the following tests:
+          //  - p4est_2d_constraintmatrix_04.*.debug
+          //  - p4est_3d_constraintmatrix_03.*.debug
+          it_input->reinit(dofs_per_cell, /*omit_zeroing_entries=*/true);
```

Comment:
This seems like a bug. Let's go with this for the moment, but please open a github issue that we look into this some more and potentially take this out.